No response sent to client when subscribing to non-existent resource

Suppose client subscribes to a resource that doesn't exist, like
```
{"jsonrpc":"2.0","id":1,"method":"resources/subscribe","params":{"uri":"file:///WRONG"}}
```
The server errors out with:
```
DEBUG: [ERROR] 2026-04-01 09:21:57,213 ERROR [io.quarkus.vertx.core.runtime.VertxCoreRecorder] (vert.x-eventloop-thread-0) Uncaught exception received by Vert.x: io.quarkiverse.mcp.server.McpException: Resource not found: file:///WRONG
```

and no response is sent to the client. IIUC it should send an `-32002` error as described in https://modelcontextprotocol.io/specification/2025-11-25/server/resources#error-handling

--- a/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ResourceManagerImpl.java
+++ b/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ResourceManagerImpl.java
@@ -121,6 +121,14 @@ public class ResourceManagerImpl extends FeatureManagerBase<ResourceResponse, Re
                 .collect(Collectors.toCollection(CopyOnWriteArrayList::new)));
     }
 
+    void unsubscribe(String uri, McpRequest mcpRequest) {
+        ResourceInfo info = getResource(uri);
+        if (info == null || !matchesServer(info, mcpRequest)) {
+            throw notFound(uri);
+        }
+        unsubscribe(uri, mcpRequest.connection().id());
+    }
+
     void unsubscribe(String uri, String connectionId) {
         List<String> ids = uriToSubscribers.get(uri);
         if (ids != null) {

--- a/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ResourceMessageHandler.java
+++ b/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ResourceMessageHandler.java
@@ -37,7 +37,11 @@ class ResourceMessageHandler extends MessageHandler {
             return mcpRequest.sender().sendError(id, JsonRpcErrorCodes.INVALID_PARAMS, "Resource URI not defined");
         }
         LOG.debugf("Subscribe to resource %s [id: %s]", resourceUri, id);
-        manager.subscribe(resourceUri, mcpRequest);
+        try {
+            manager.subscribe(resourceUri, mcpRequest);
+        } catch (McpException e) {
+            return mcpRequest.sender().sendError(id, e.getJsonRpcErrorCode(), e.getMessage());
+        }
         // Send empty result
         return mcpRequest.sender().sendResult(id, new JsonObject());
     }
@@ -50,7 +54,11 @@ class ResourceMessageHandler extends MessageHandler {
             return mcpRequest.sender().sendError(id, JsonRpcErrorCodes.INVALID_PARAMS, "Resource URI not defined");
         }
         LOG.debugf("Unsubscribe to resource %s [id: %s]", resourceUri, id);
-        manager.unsubscribe(resourceUri, mcpRequest.connection().id());
+        try {
+            manager.unsubscribe(resourceUri, mcpRequest);
+        } catch (McpException e) {
+            return mcpRequest.sender().sendError(id, e.getJsonRpcErrorCode(), e.getMessage());
+        }
         // Send empty result
         return mcpRequest.sender().sendResult(id, new JsonObject());
     }

--- a/transports/http/deployment/src/test/java/io/quarkiverse/mcp/server/test/resources/InvalidResourceSubscribeTest.java
+++ b/transports/http/deployment/src/test/java/io/quarkiverse/mcp/server/test/resources/InvalidResourceSubscribeTest.java
@@ -1,0 +1,51 @@
+package io.quarkiverse.mcp.server.test.resources;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkiverse.mcp.server.JsonRpcErrorCodes;
+import io.quarkiverse.mcp.server.McpMethod;
+import io.quarkiverse.mcp.server.runtime.Messages;
+import io.quarkiverse.mcp.server.test.McpAssured;
+import io.quarkiverse.mcp.server.test.McpAssured.McpStreamableTestClient;
+import io.quarkiverse.mcp.server.test.McpServerTest;
+import io.quarkus.test.QuarkusUnitTest;
+import io.vertx.core.json.JsonObject;
+
+public class InvalidResourceSubscribeTest extends McpServerTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = defaultConfig()
+            .withEmptyApplication();
+
+    @Test
+    public void testSubscribeToNonExistentResource() {
+        McpStreamableTestClient client = McpAssured.newConnectedStreamableClient();
+        client.when()
+                .message(Messages.newRequest(100, McpMethod.RESOURCES_SUBSCRIBE.jsonRpcName(),
+                        new JsonObject().put("uri", "file:///nonexistent")))
+                .withErrorAssert(error -> {
+                    assertEquals(JsonRpcErrorCodes.RESOURCE_NOT_FOUND, error.code());
+                    assertEquals("Resource not found: file:///nonexistent", error.message());
+                })
+                .send()
+                .thenAssertResults();
+    }
+
+    @Test
+    public void testUnsubscribeFromNonExistentResource() {
+        McpStreamableTestClient client = McpAssured.newConnectedStreamableClient();
+        client.when()
+                .message(Messages.newRequest(101, McpMethod.RESOURCES_UNSUBSCRIBE.jsonRpcName(),
+                        new JsonObject().put("uri", "file:///nonexistent")))
+                .withErrorAssert(error -> {
+                    assertEquals(JsonRpcErrorCodes.RESOURCE_NOT_FOUND, error.code());
+                    assertEquals("Resource not found: file:///nonexistent", error.message());
+                })
+                .send()
+                .thenAssertResults();
+    }
+
+}

